papers/libceed-paper/paper.rst

```diff
@@ -0,0 +1,72 @@
+stability properties offered by higher order discretizations. Nowadays, high-order numerical methods are widely used in Partial Differential Equation (PDE) solvers, but software packages that provide high-performance implementations have often been special-purpose and intrusive. In contrast, libCEED is light-weight, minimally intrusive, and very versatile. When high-order finite/spectral element discretizations are used, the resulting sparse matrix representation of a global operator is computationally expensive, with respect to both the memory transfer and floating point operations needed for its evaluation. libCEED's Application Programming Interface (API) provides the local action of an operator (linear or nonlinear) without assembling its sparse representation. The purely algebraic nature of libCEED allows efficient operator evaluations (selectable at runtime) and offers matrix-free preconditioning ingredients. While libCEED’s focus is on high-order finite elements, the approach with which it is designed is algebraic and thus applicable to other discretizations in factored form.
+
+.. figure:: TensorVsAssembly.png
+   :align: center
+   :figclass: bht
+
+   Comparison of memory transfer and floating point operations per degree of freedom for different representations of a linear operator for a PDE in 3D with :math:`b` components and variable coefficients arising due to Newton linearization of a material nonlinearity. The representation labeled as *tensor* computes metric terms on the fly and stores a compact representation of the linearization at quadrature points. The representation labeled as *tensor-qstore* pulls the metric terms into the stored representation. The *assembled* representation uses a (block) CSR format.
+
+In this work, we first introduce libCEED’s conceptual framework and C interface, and then illustrate its new Python interface, developed using the C Foreign Function Interface (CFFI) for Python. CFFI allows to reuse most of the C declarations and requires only a minimal adaptation of some of them. The C and Python APIs are mapped in a nearly 1:1 correspondence. For instance, data stored in the CeedVector structure are associated to arrays defined via the numpy package. In fact, since libCEED heavily relies on pointers and arrays to handle the data, a Python structure that resembles the C arrays is needed. In details, numpy arrays allow this correspondence obtained by passing the numpy array memory address as pointers to the libCEED C API.
+
+libCEED's API
+----------------------------------------------------------------------------------
+
+LibCEED's API provides the local action of the linear or nonlinear operator without assembling its sparse representation. Let us define the global operator as
```

Comment:
I find the word "local" confusing here, I would reword like this:
"LibCEED's API provides the action of the linear or nonlinear Finite Element operator without assembling its sparse representation."

Reply:
I think it would be a bit better to give the intuition that instead of forming a single operator using a sparse matrix, we're going to use a product of operations to decompose our operator that will be applied __sequentially__.

Reply:
Attempted in [8187f04](https://github.com/valeriabarra/scipy_proceedings/pull/1/commits/8187f0446ea4ca87e782424c4e7d6d0514e9a246) please let me know if it sounds good

Reply:
We perhaps don't want to say the word sequentially? The exact implementation is backend specific. In some cases we apply suboperators partially, like in the CPU backends. What about:

Instead of forming a single operator using a sparse matrix representation, libCEED composes the subcomponents described in equation (:ref:`eq-operator-decomposition`) to provide the action of the operator in a fashion tuned for the hardware available to the user.

Or something like that?

Reply:
Attempted again in [204b47d](https://github.com/valeriabarra/scipy_proceedings/pull/1/commits/204b47d14efbdf3e04cfdcd0f957a72c2c855eec)

Reply:
I like it

Reply:
All the implementations are sequential, but are more or less local.

Reply:
Sort of?
In the blocked CPU backends, some restrictions are fully applied and some are partially applied.
Then in a loop over element batches, active restrictions are applied to a portion of the inputs, the basis operations are applied to a portion of the elements, qfunction is applied for a portion of the elements, the output basis operations are applied to a portion of the elements, and output restrictions are applied to a portion of the elements.

I personally wouldn't call that a sequential application of objects in the decomposition in the way the that all restrictions, all basis operations, a qfunction call, all transpose operations, and all transpose restrictions are applied fully in sequence in non-optimized backends.

Reply:
I wrote "more or less local" the operations are still applied sequentially...